tools/sync-fault-leds.hpp

```diff
@@ -10,14 +11,186 @@
-    //    property from LED.Controller.pca955x_* service.
-    //    If State property is incorrect, update it.
+    try
+    {
+        // extract the FRU name from object path
+        const auto l_lastSlashPos{i_objectPath.find_last_of('/')};
+        if (l_lastSlashPos == std::string::npos)
+        {
+            throw std::runtime_error("Invalid object path");
+        }
+
+        const std::string l_fruName = i_objectPath.substr(l_lastSlashPos + 1);
+        std::cout << "FRU name: " << l_fruName << std::endl;
+
+        // lambda to get the Functional property of the FRU
+        auto l_getFunctionalProperty =
+            [&i_objectPath =
+                 std::as_const(i_objectPath)]() -> std::optional<bool> {
+            std::optional<bool> l_result;
+
+            const auto l_functionalProperty = utility::getProperty(
+                "xyz.openbmc_project.Inventory.Manager",
+                "/xyz/openbmc_project/inventory" + i_objectPath,
+                "xyz.openbmc_project.State.Decorator.OperationalStatus",
+                "Functional");
+
+            if (const auto l_functionalPropertyVal =
+                    std::get_if<bool>(&l_functionalProperty))
+            {
+                l_result.emplace(*l_functionalPropertyVal);
+            }
+            return l_result;
+        };
+
+        // lambda to get the Asserted property of the fault LED
+        auto l_getAssertedProperty =
+            [](const std::string& i_fruName) -> std::optional<bool> {
+            std::optional<bool> l_result;
+
+            const auto l_assertedProperty = utility::getProperty(
+                "xyz.openbmc_project.LED.GroupManager",
+                "/xyz/openbmc_project/led/groups/" + i_fruName + "_fault",
+                "xyz.openbmc_project.Led.Group", "Asserted");
+
+            if (const auto l_assertedPropertyValue =
+                    std::get_if<bool>(&l_assertedProperty))
+            {
+                l_result.emplace(*l_assertedPropertyValue);
+            }
+            return l_result;
+        };
+
+        // lambda to get the physical LED state
+        auto l_getPhysicalLedState =
+            [](const std::string& i_fruName) -> std::optional<std::string> {
+            std::optional<std::string> l_result;
+
+            const auto l_assertedProperty = utility::getProperty(
+                "xyz.openbmc_project.LED.Controller.pca955x_" + i_fruName,
+                "/xyz/openbmc_project/led/physical/pca955x_" + i_fruName,
+                "xyz.openbmc_project.Led.Physical", "State");
+
+            if (const auto l_assertedPropertyValue =
+                    std::get_if<std::string>(&l_assertedProperty))
+            {
+                l_result.emplace(*l_assertedPropertyValue);
+            }
+            return l_result;
+        };
+
+        // lambda to set the Asserted property of the fault LED
+        auto l_setAssertedProperty = [](const std::string& i_fruName,
```

Comment:
I don't see a need of defining lambdas at so many places if we are calling it only once. Why can't we call all these functions in place and avoid unnecessary coding?